Firestore/core/src/firebase/firestore/local/leveldb_persistence_posix.cc

```diff
@@ -0,0 +1,49 @@
+/*
+ * Copyright 2019 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "Firestore/core/src/firebase/firestore/local/leveldb_persistence.h"
+
+#include "Firestore/core/src/firebase/firestore/util/path.h"
+#include "Firestore/core/src/firebase/firestore/util/status.h"
+
+#if !defined(__APPLE__) && !defined(_WIN32)
+
+#include <pwd.h>
+#include <sys/types.h>
+#include <unistd.h>
+
+namespace firebase {
+namespace firestore {
+namespace local {
+
+using util::Path;
+
+Path LevelDbPersistence::AppDataDirectory() {
+  const char* home_dir;
+
+  if ((home_dir = getenv("HOME")) == NULL) {
```

Comment:
If we're running on Linux (but not Android), we really should follow the [XDG Base Directory Specification](https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html) for user specific data files. That is:

  * The check should be inside `#if __linux__ && !__ANDROID__`
  * getenv("XDG_DATA_HOME"); if set use it for `data_home`.
  * if unset find the home directory and then path append `.local/share` (this is what argues for `data_home` being a `Path`).

Only on non-Linux POSIX should we fall back on what you have here (though we have no supported platforms of this type there's nothing stopping someone from trying this on e.g. Solaris).

I suggest factoring out `HomeDirectory` that does the `$HOME` then fallback on `getpwuid` so you can reuse that in both Linux and non-Linux cases.